lib/optimize/ConcatenatedModule.js

```diff
@@ -0,0 +1,489 @@
+	}
+
+	findNewName(oldName, usedNamed, extraInfo) {
+		let name = oldName;
+
+		if(name === "__WEBPACK_MODULE_DEFAULT_EXPORT__")
+			name = "defaultExport";
+
+		const splittedInfo = extraInfo.split("/");
+		while(splittedInfo.length) {
+			name = splittedInfo.pop() + "_" + name;
+			if(!usedNamed.has(Template.toIdentifier(name))) return Template.toIdentifier(name);
+		}
+
+		while(!usedNamed.has(name = name + "_"));
+		return name;
+	}
+
+}
+
+class HarmonyImportSpecifierDependencyConcatenatedTemplate {
+	constructor(originalTemplate, modulesMap) {
+		this.originalTemplate = originalTemplate;
+		this.modulesMap = modulesMap;
+	}
+
+	apply(dep, source, outputOptions, requestShortener, dependencyTemplates) {
+		const module = dep.importDependency.module;
+		const info = this.modulesMap.get(module);
+		if(!info) {
+			this.originalTemplate.apply(dep, source, outputOptions, requestShortener, dependencyTemplates);
+			return;
+		}
+		if(!Array.isArray(module.providedExports))
+			throw new Error(`Module ${module.resource} has no static exports ${module.providedExports}`);
+		let content;
+		if(dep.id === null) {
+			content = `__WEBPACK_MODULE_REFERENCE__${info.index}_ns__`;
+		} else {
+			const exportIdx = (module.providedExports).indexOf(dep.id);
+			content = exportIdx === -1 ? "undefined" : `__WEBPACK_MODULE_REFERENCE__${info.index}_${exportIdx}__`;
+		}
+		if(dep.shorthand) {
+			content = dep.name + ": " + content;
+		}
+		source.replace(dep.range[0], dep.range[1] - 1, content);
+	}
+}
+
+class HarmonyImportDependencyConcatenatedTemplate {
+	constructor(originalTemplate, modulesMap) {
+		this.originalTemplate = originalTemplate;
+		this.modulesMap = modulesMap;
+	}
+
+	apply(dep, source, outputOptions, requestShortener, dependencyTemplates) {
+		const module = dep.module;
+		const info = this.modulesMap.get(module);
+		if(!info) {
+			this.originalTemplate.apply(dep, source, outputOptions, requestShortener, dependencyTemplates);
+			return;
+		}
+		source.replace(dep.range[0], dep.range[1] - 1, "");
+	}
+}
+
+class HarmonyExportSpecifierDependencyConcatenatedTemplate {
+	constructor(originalTemplate, rootModule) {
+		this.originalTemplate = originalTemplate;
+		this.rootModule = rootModule;
+	}
+
+	apply(dep, source, outputOptions, requestShortener, dependencyTemplates) {
+		if(dep.originModule === this.rootModule) {
+			this.originalTemplate.apply(dep, source, outputOptions, requestShortener, dependencyTemplates);
+		}
+	}
+}
+
+class HarmonyExportExpressionDependencyConcatenatedTemplate {
+	constructor(originalTemplate, rootModule) {
+		this.originalTemplate = originalTemplate;
+		this.rootModule = rootModule;
+	}
+
+	apply(dep, source, outputOptions, requestShortener, dependencyTemplates) {
+		if(dep.originModule === this.rootModule) {
+			this.originalTemplate.apply(dep, source, outputOptions, requestShortener, dependencyTemplates);
+		} else {
+			const content = "/* harmony default export */ var __WEBPACK_MODULE_DEFAULT_EXPORT__ = ";
```

Comment:
So the name is constant, but it's replaced in the module before the concatenation step, at which point it gets renamed to something derived from the module name + `defaultExport`, right?